Use root obligation on E0277 for some cases

estebank opened the pull request:
When encountering trait bound errors that satisfy some heuristics that tell us that the relevant trait for the user comes from the root obligation and not the current obligation, we use the root predicate for the main message.

This allows to talk about "X doesn't implement Pattern<'_>" over the most specific case that just happened to fail, like  "char doesn't implement Fn(&mut char)" in
`tests/ui/traits/suggest-dereferences/root-obligation.rs`

The heuristics are:

 - the type of the leaf predicate is (roughly) the same as the type from the root predicate, as a proxy for "we care about the root"
 - the leaf trait and the root trait are different, so as to avoid talking about `&mut T: Trait` and instead remain talking about `T: Trait` instead
 - the root trait is not `Unsize`, as to avoid talking about it in `tests/ui/coercion/coerce-issue-49593-box-never.rs`.

```
error[E0277]: the trait bound `&char: Pattern<'_>` is not satisfied
  --> $DIR/root-obligation.rs:6:38
   |
LL |         .filter(|c| "aeiou".contains(c))
   |                             -------- ^ the trait `Fn<(char,)>` is not implemented for `&char`, which is required by `&char: Pattern<'_>`
   |                             |
   |                             required by a bound introduced by this call
   |
   = note: required for `&char` to implement `FnOnce<(char,)>`
   = note: required for `&char` to implement `Pattern<'_>`
note: required by a bound in `core::str::<impl str>::contains`
  --> $SRC_DIR/core/src/str/mod.rs:LL:COL
help: consider dereferencing here
   |
LL |         .filter(|c| "aeiou".contains(*c))
   |                                      +
```

Fix #79359, fix #119983, fix #118779, cc #118415 (the suggestion needs to change), cc #121398 (doesn't fix the underlying issue).

estebank committed:
Use root obligation on E0277 for some cases

When encountering trait bound errors that satisfy some heuristics that
tell us that the relevant trait for the user comes from the root
obligation and not the current obligation, we use the root predicate for
the main message.

This allows to talk about "X doesn't implement Pattern<'_>" over the
most specific case that just happened to fail, like  "char doesn't
implement Fn(&mut char)" in
`tests/ui/traits/suggest-dereferences/root-obligation.rs`

The heuristics are:

 - the type of the leaf predicate is (roughly) the same as the type
   from the root predicate, as a proxy for "we care about the root"
 - the leaf trait and the root trait are different, so as to avoid
   talking about `&mut T: Trait` and instead remain talking about
   `T: Trait` instead
 - the root trait is not `Unsize`, as to avoid talking about it in
   `tests/ui/coercion/coerce-issue-49593-box-never.rs`.

```
error[E0277]: the trait bound `&char: Pattern<'_>` is not satisfied
  --> $DIR/root-obligation.rs:6:38
   |
LL |         .filter(|c| "aeiou".contains(c))
   |                             -------- ^ the trait `Fn<(char,)>` is not implemented for `&char`, which is required by `&char: Pattern<'_>`
   |                             |
   |                             required by a bound introduced by this call
   |
   = note: required for `&char` to implement `FnOnce<(char,)>`
   = note: required for `&char` to implement `Pattern<'_>`
note: required by a bound in `core::str::<impl str>::contains`
  --> $SRC_DIR/core/src/str/mod.rs:LL:COL
help: consider dereferencing here
   |
LL |         .filter(|c| "aeiou".contains(*c))
   |                                      +
```

Fix #79359, fix #119983, fix #118779, cc #118415 (the suggestion needs
to change).

estebank commented:
CC @lcnr, @compiler-errors, @oli-obk & @fmease. This is an attempt to mitigate the issue highlighted at https://rust-lang.zulipchat.com/#narrow/stream/144729-t-types/topic/Change.20to.20default.20E0277.20message and the linked tickets. There are many more things to be fixed in these errors, but this seemed like a small enough change that could pull its weight to stop confusing people in cases like the `T: Pattern<'_>` one.

oli-obk commented on library/core/src/iter/traits/collect.rs:
````diff
@@ -236,6 +236,49 @@ pub trait FromIterator<A>: Sized {
 /// ```
 #[rustc_diagnostic_item = "IntoIterator"]
 #[rustc_skip_array_during_method_dispatch]
+#[rustc_on_unimplemented(
````
are these copied from iterator or is there a difference? Maybe document that with a normal comment

estebank replied:
Yes, I took them straight from Iterator and Future. I can iterate on whether removing the others is needed or if only a subset of conditions are needed. I strongly suspect we need to keep all on Iterator and only need a subset on IntoIterator

estebank replied:
So I took a quick look and funnily enough this change actually improves the situation of not knowing when the error is coming from an expression desugaring or type context, because all desugaring uses the `Into*` traits, while most normal code uses the actual traits, so we can leave the expression level suggestions in the conversion traits (almost, deref would need to be accounted, as per the `FIXME` comment) exclusively.

oli-obk commented on compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:
```diff
@@ -409,9 +409,44 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
+                            // We rely on a few heuristics to identify cases where this root
+                            // obligation is more important than the leaf obligation:
+                            //
+                            //  - the type of the leaf predicate is (roughly) the same as the type
+                            //    from the root predicate, as a proxy for "we care about the root"
+                            //  - the leaf trait and the root trait are different, so as to avoid
+                            //    talking about `&mut T: Trait` and instead remain talking about
+                            //    `T: Trait` instead
+                            //  - the root trait is not `Unsize`, as to avoid talking about it in
+                            //    `tests/ui/coercion/coerce-issue-49593-box-never.rs`.
```
inline these heuristics with their actual checks above, to help keep them in sync with future changes

estebank replied:
Will do. I prefer that style as well, I've moved away from it recently because it messes up with rustfmt.

estebank commented on tests/ui/traits/suggest-dereferences/root-obligation.stderr:
```diff
@@ -1,12 +1,11 @@
-error[E0277]: expected a `Fn(char)` closure, found `char`
+error[E0277]: the trait bound `&char: Pattern<'_>` is not satisfied
```
This error, in the future (not this PR) it should actually be:

```
error[E0277]: the trait bound `&char: Pattern<'_>` is not satisfied
  --> $DIR/root-obligation.rs:6:38
   |
LL |         .filter(|c| "aeiou".contains(c))
   |                             -------- ^ the trait `Pattern<'_>` is not implemented for `&char`
   |                             |
   |                             required by a bound introduced by this call
   |
note: required by a bound in `core::str::contains`
  --> $SRC_DIR/core/src/str/mod.rs:LL:COL
help: consider dereferencing here
   |
LL |         .filter(|c| "aeiou".contains(*c))
   |                                      +
```

instead of

```
error[E0277]: the trait bound `&char: Pattern<'_>` is not satisfied
  --> $DIR/root-obligation.rs:6:38
   |
LL |         .filter(|c| "aeiou".contains(c))
   |                             -------- ^ the trait `Fn<(char,)>` is not implemented for `&char`, which is required by `&char: Pattern<'_>`
   |                             |
   |                             required by a bound introduced by this call
   |
   = note: required for `&char` to implement `FnOnce<(char,)>`
   = note: required for `&char` to implement `Pattern<'_>`
note: required by a bound in `core::str::<impl str>::contains`
  --> $SRC_DIR/core/src/str/mod.rs:LL:COL
help: consider dereferencing here
   |
LL |         .filter(|c| "aeiou".contains(*c))
   |                                      +
```

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8115731045/job/22184197035) [(plain)](https://github.com/rust-lang/rust/commit/ccb8b8a8809c9cc9782b520b7791f39afd40396a/checks/22184197035/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#10 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#10 DONE 0.0s

#11 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#11 0.609   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#11 0.626 Collecting boolean-py==4.0
#11 0.633   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#11 0.650 Collecting chardet==5.1.0
---
#11 3.714 Building wheels for collected packages: reuse
#11 3.715   Building wheel for reuse (pyproject.toml): started
#11 4.044   Building wheel for reuse (pyproject.toml): finished with status 'done'
#11 4.045   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#11 4.045   Stored in directory: /tmp/pip-ephem-wheel-cache-4tyoybzb/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#11 4.047 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#11 4.069   Attempting uninstall: setuptools
#11 4.070     Found existing installation: setuptools 59.6.0
#11 4.071     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#11 4.071     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#11 4.071     Can't uninstall 'setuptools'. No files were found to uninstall.
#11 4.747 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#11 4.748 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#11 5.261 Collecting virtualenv
#11 5.324   Downloading virtualenv-20.25.1-py3-none-any.whl (3.8 MB)
#11 5.496      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 22.1 MB/s eta 0:00:00
#11 5.543 Collecting distlib<1,>=0.3.7
#11 5.550   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#11 5.596 Collecting filelock<4,>=3.12.2
#11 5.603   Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
#11 5.603   Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
#11 5.635 Collecting platformdirs<5,>=3.9.1
#11 5.643   Downloading platformdirs-4.2.0-py3-none-any.whl (17 kB)
#11 5.727 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#11 5.888 Successfully installed distlib-0.3.8 filelock-3.13.1 platformdirs-4.2.0 virtualenv-20.25.1
#11 DONE 6.0s

#12 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#12 DONE 0.0s
---
DirectMap4k:      192448 kB
DirectMap2M:     7147520 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6db96de66c2c0ea3f4f2f348ed1a83c2c507687d/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6db96de66c2c0ea3f4f2f348ed1a83c2c507687d-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 27.54s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/core/src/future/future.rs at line 30:
 #[lang = "future_trait"]
 #[diagnostic::on_unimplemented(
     label = "`{Self}` is not a future",
-    message = "`{Self}` is not a future",
+    message = "`{Self}` is not a future"
 pub trait Future {
     /// The type of value produced on completion.
     /// The type of value produced on completion.
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/default.rs" "/checkout/library/core/src/prelude/v1.rs" "/checkout/library/core/src/prelude/mod.rs" "/checkout/library/core/src/future/into_future.rs" "/checkout/library/core/src/future/join.rs" "/checkout/library/core/src/future/mod.rs" "/checkout/library/core/src/future/ready.rs" "/checkout/library/core/src/future/future.rs" "/checkout/library/core/src/future/pending.rs" "/checkout/library/core/src/future/poll_fn.rs" "/checkout/library/core/src/async_iter/mod.rs" "/checkout/library/core/src/async_iter/async_iter.rs" "/checkout/library/core/src/async_iter/from_iter.rs" "/checkout/library/core/src/hint.rs" "/checkout/library/core/src/panic.rs" "/checkout/library/core/src/cell/lazy.rs" "/checkout/library/core/src/cell/once.rs" "/checkout/library/core/src/time.rs" "/checkout/library/core/src/ffi/mod.rs" "/checkout/library/core/src/ffi/c_str.rs" "/checkout/library/core/src/clone.rs" "/checkout/library/core/src/internal_macros.rs" "/checkout/library/core/src/ascii.rs" "/checkout/library/core/src/intrinsics.rs" "/checkout/library/core/src/io/mod.rs" "/checkout/library/core/src/io/borrowed_buf.rs" "/checkout/library/core/src/str/iter.rs" "/checkout/library/core/src/str/validations.rs" "/checkout/library/core/src/str/count.rs" "/checkout/library/core/src/str/error.rs" "/checkout/library/core/src/str/pattern.rs" "/checkout/library/core/src/str/lossy.rs" "/checkout/library/core/src/str/converts.rs" "/checkout/library/core/src/str/traits.rs" "/checkout/library/core/src/str/mod.rs" "/checkout/library/core/src/hash/sip.rs" "/checkout/library/core/src/hash/mod.rs" "/checkout/library/core/src/escape.rs" "/checkout/library/core/src/char/mod.rs" "/checkout/library/core/src/char/decode.rs" "/checkout/library/core/src/char/convert.rs" "/checkout/library/core/src/char/methods.rs" "/checkout/library/core/src/intrinsics/mir.rs" "/checkout/library/core/src/intrinsics/simd.rs" "/checkout/library/core/src/unit.rs" "/checkout/library/core/src/bool.rs" "/checkout/library/core/src/panicking.rs" "/checkout/library/core/src/primitive_docs.rs" "/checkout/library/core/src/cell.rs" "/checkout/library/core/src/macros/mod.rs" "/checkout/library/core/src/convert/num.rs" "/checkout/library/core/src/convert/mod.rs" "/checkout/library/core/src/ascii/ascii_char.rs" "/checkout/library/core/src/borrow.rs" "/checkout/library/core/src/unicode/mod.rs" "/checkout/library/core/src/unicode/unicode_data.rs" "/checkout/library/core/src/unicode/printable.rs" "/checkout/library/core/src/arch.rs" "/checkout/library/core/src/pin.rs" "/checkout/library/core/src/any.rs" "/checkout/library/core/src/cmp/bytewise.rs" "/checkout/library/core/tests/tuple.rs" "/checkout/library/core/tests/result.rs" "/checkout/library/core/src/option.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri Mar  1 18:41:07 UTC 2024
  network time: Fri, 01 Mar 2024 18:41:07 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

estebank committed:
Be more lax in `.into_iter()` suggestion when encountering `Iterator` methods on non-`Iterator`

```
error[E0599]: no method named `map` found for struct `Vec<bool>` in the current scope
  --> $DIR/vec-on-unimplemented.rs:3:23
   |
LL |     vec![true, false].map(|v| !v).collect::<Vec<_>>();
   |                       ^^^ `Vec<bool>` is not an iterator
   |
help: call `.into_iter()` first
   |
LL |     vec![true, false].into_iter().map(|v| !v).collect::<Vec<_>>();
   |                       ++++++++++++
```

We used to provide some help through `rustc_on_unimplemented` on non-`impl Trait` and non-type-params, but this lets us get rid of some otherwise unnecessary conditions in the annotation on `Iterator`.

oli-obk commented on compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:
```diff
@@ -409,9 +409,51 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
+                            && (
+                                // `T: Trait` && `&&T: OtherTrait`, we want `OtherTrait`
+                                trait_predicate.self_ty().skip_binder()
+                                    == root_pred.self_ty().peel_refs()
```
is it intended that we treat types with lifetimes differently if their lifetimes differ? We'll just have fewer cases where we switch to the root obligation, but I'm not sure whether it'll miss out on cases where we'd want to use the root.

I'll give it a more thorough look next week, but if you have ideas here, maybe add some tests where the self type has lifetimes

estebank replied:
We could use `can_eq` instead here.

oli-obk replied:
Yea good point. If that shows some tests that are affected, good, otherwise we need to write some

estebank replied:
None of new tests were affected (after I checked for no escaping vars). I'll have to think of a case where lifetime divergence is the issue *but* that doesn't manifest as an `OutlivesPredicate` clause.

estebank committed:
Use `can_eq` instead of `Ty<'_> == Ty<'_>`

fmease commented on tests/ui/associated-types/substs-ppaux.normal.stderr:
```diff
@@ -70,13 +70,12 @@ help: use parentheses to call this function
    |
 LL |     <str as Foo<u8>>::bar;
-   |      ^^^ doesn't have a size known at compile-time
+   |      ^^^ the trait `Sized` is not implemented for `str`, which is required by `str: Foo<'_, '_, u8>`
```
I haven't read the implementation or the most recent Zulip comment and just saw this stderr diff, so sorry if this has been discussed already.

Is this intentional? Shouldn't we still use the `[rustc_on_unimplemented]`/`#[diagnostic::on_unimplemented]` message here despite the derived/root obligation changes?

estebank replied:
It is somewhat intentional. When you have `T: Sized` then you want `T` to have a compile time known size. When you have `T: Foo, Foo: Sized`, I think it makes sense to go back to "`T` does not implement `Sized`", it's a subtle difference of "what does the user care about".

estebank replied:
Also, I do want to do a full refactor of this logic, too many ad-hoc hacks have piled beyond maintainability (in no small part due to me), but now that we know what the *output* we want to keep and change is, it will be easier to clean this up in a way that makes sense. This PR is about putting the output piece closer to that "ideal end state" in anticipation of that work.

bors commented:
:umbrella: The latest upstream changes (presumably #121890) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 40f9dccad64c20fdcc91e5e0da15d8647465820f has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 40f9dccad64c20fdcc91e5e0da15d8647465820f -->
<!-- homu: {"type":"Approved","sha":"40f9dccad64c20fdcc91e5e0da15d8647465820f","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
